src/dict.c

```diff
@@ -123,133 +130,156 @@ int _dictInit(dict *d, dictType *type)
- * guaranteed that this function will rehash even a single bucket, since it
- * will visit at max N*10 empty buckets in total, otherwise the amount of
- * work it does would be unbound and the function may block for a long time. */
-int dictRehash(dict *d, int n) {
-    int empty_visits = n*10; /* Max number of empty buckets to visit. */
+ * Note that a rehashing step consists in updating a bucket (that may have more
+ * than one key as we use chaining). However, since part of the hash table may be
+ * composed of empty buckets, it is not guaranteed that this function will rehash
+ * even a single bucket, since it will visit at max bucket_visits*10 empty
+ * buckets in total, otherwise the amount of work it does would be unbound and
+ * the function may block for a long time. */
+int dictRehash(dict *d, int bucket_visits) {
+    /* Max number of empty buckets to visit. */
+    int empty_visits = bucket_visits * 10;
+
     if (!dictIsRehashing(d)) return 0;
 
-    while(n-- && d->ht_used[0] != 0) {
-        dictEntry *de, *nextde;
+    /* Iterate and rehash all buckets up to the previous size of the hash table,
+     * and rehash all entries in buckets to their updated buckets */
+    long ht_size = DICTHT_SIZE(d->ht_size_exp);
+    long ht_size_prev = DICTHT_SIZE(d->ht_size_exp_prev);
+
+    while (bucket_visits > 0) {
+        dictEntry *entries_to_rehash = d->ht_table[d->rehashidx];
 
-        /* Note that rehashidx can't overflow as we are sure there are more
-         * elements because ht[0].used != 0 */
-        assert(DICTHT_SIZE(d->ht_size_exp[0]) > (unsigned long)d->rehashidx);
-        while(d->ht_table[0][d->rehashidx] == NULL) {
-            d->rehashidx++;
+        /* If HT grows, need to set to NULL current bucket and corresponding new ones
+         * before rehashing back entries (as we avoid from using memset when
+         * expanded the table). If shrinks this loop iterates once and set to NULL
+         * current bucket.  */
+        long idx = d->rehashidx ;
+        do {
+            d->ht_table[idx] = NULL;
+            idx += ht_size_prev;
+        } while (idx < ht_size);
+
+        if (entries_to_rehash == NULL) {
             if (--empty_visits == 0) return 1;
+        } else {
+            --bucket_visits;
+            dictEntry *de = entries_to_rehash;
+
+            /* rehash all the keys in current bucket */
+            while (de) {
+                long h;
+
+                dictEntry *nextde = de->next;
+                /* Get the new index based on the new resized hash table.
+                 * Note that if HT shrinks, this operation will take care to
+                 * rehash to corresponding bucket (by dropping MSB bit(s)) */
+                h = dictHashKey(d, de->key) & DICTHT_SIZE_MASK(d->ht_size_exp);
+
+                /* Chain entry to the new bucket */
+                de->next = d->ht_table[h];
+                d->ht_table[h] = de;
+                de = nextde;
+            }
```

Comment:
I think of a situation, for example, before the expansion is 8 slots, after the expansion is 16 slots, if an entry is at 3 slots, if it is rehashed to 4 slots, then when the rehashidx to 4, this element will be rehashed again?

Reply:
Cannot happen. On expansion, if some entry actually moved to another bucket, it always will be new allocated bucket, with index behind the size of the old hash table. And rehashing iterates up-to the size of the old table. 

Continue your example with resize from 8 to 16, each entry with address [XXX], X={0,1}, at the end of the rehash, will either stay at its place, or moved to address 1[XXX] which is a new allocated buffer with index behind the size of the old table.

Reply:
I thought the same when I saw the code for the first time :)